io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs

```diff
@@ -535,6 +544,51 @@ class MonadInspectSTM m
                       -> m ()
   traceTSemIO = \v f -> atomically (traceTSem Proxy v f)
 
+traceTVarShow :: (MonadTraceSTM m, Show a)
+               => proxy m
+               -> TVar m a
+               -> STM m ()
+traceTVarShow p tvar =
+  traceTVar p tvar (\pv v -> pure $ TraceString $ case (pv, v) of
+          (Nothing, st') -> "Created: " <> show st'
+          (Just st', st'') -> "Modified: " <> show st' <> " -> " <> show st''
+      )
+
+traceTVarShowIO :: (MonadTraceSTM m, Show a)
+               => TVar m a
+               -> m ()
+traceTVarShowIO tvar =
+  traceTVarIO tvar (\pv v -> pure $ TraceString $ case (pv, v) of
+          (Nothing, st') -> "Created: " <> show st'
+          (Just st', st'') -> "Modified: " <> show st' <> " -> " <> show st''
+      )
+
+traceTMVarShow :: (MonadTraceSTM m, Show a)
+               => proxy m
+               -> TMVar m a
+               -> STM m ()
+traceTMVarShow p tmvar =
+  traceTMVar p tmvar (\pv v -> pure $ TraceString $ case (pv, v) of
+          (Nothing, Nothing) -> "Created empty"
+          (Nothing, Just st') -> "Created full: " <> show st'
+          (Just Nothing, Just st') -> "Put: " <> show st'
+          (Just Nothing, Nothing) -> "Remains empty"
+          (Just Just{}, Nothing) -> "Take"
+          (Just (Just st'), Just st'') -> "Modified: " <> show st' <> " -> " <> show st''
+      )
+
+traceTMVarShowIO :: (Show a, MonadTraceSTM m)
+                 => TMVar m a
+                 -> m ()
+traceTMVarShowIO tmvar =
+  traceTMVarIO tmvar (\pv v -> pure $ TraceString $ case (pv, v) of
+          (Nothing, Nothing) -> "Created empty"
+          (Nothing, Just st') -> "Created full: " <> show st'
+          (Just Nothing, Just st') -> "Put: " <> show st'
+          (Just Nothing, Nothing) -> "Remains empty"
+          (Just Just{}, Nothing) -> "Take"
+          (Just (Just st'), Just st'') -> "Modified: " <> show st' <> " -> " <> show st''
+      )
```

Comment:
Maybe a better name for these functions would be `debugTVar,debugTVarIO`?

Reply:
Sure thing!